### PR TITLE
Properly checks struct members in alias expression

### PR DESCRIPTION
**Pull request by feliperodri**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/contracts/assigns_enforce_arrays_02/test.desc [hunk 1]
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
Comment (martin-cs) [inline regression/contracts/assigns_enforce_arrays_02/test.desc]: A. If this was working before, has this PR broken it?
B. If it wasn't then why did the test pass?

File: regression/contracts/assigns_type_checking_valid_cases/test.desc [hunk 2]
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
Comment (martin-cs) [inline regression/contracts/assigns_type_checking_valid_cases/test.desc]: Same question again.

**Review by martin-cs — COMMENTED** [review]
The code looks fairly good but I am worried about the tests and about disabling so many of them in a PR claiming to fix things.

File: regression/contracts/assigns_enforce_arrays_02/test.desc (hunk 1 above)
Comment (feliperodri) [reply regression/contracts/assigns_enforce_arrays_02/test.desc]: Fixed bug.

File: regression/contracts/assigns_type_checking_valid_cases/test.desc (hunk 2 above)
Comment (feliperodri) [reply regression/contracts/assigns_type_checking_valid_cases/test.desc]: Fixed bug.

**Comment by feliperodri** [comment]
> The code looks fairly good but I am worried about the tests and about disabling so many of them in a PR claiming to fix things.

@martin-cs could you take another look? I fixed the bug when dealing with arrays.

**Review by martin-cs — APPROVED** [review]
Thank you; that is a massive improvement on the previous version.

File: src/goto-instrument/contracts/assigns.cpp [hunk 3]
```diff
@@ -56,25 +55,64 @@ assigns_clause_targett::~assigns_clause_targett()
...
+  // If assigns target was a dereference, comparing objects is enough
+  if(target_id == ID_dereference)
+  {
+    return conjunction(condition);
+  }
+
+  if(target_id == ID_dereference)
+  {
+    return conjunction(condition);
+  }
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]: Remove the 4 lines above? They look repeated

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
I am still thinking about the new conditions we have added, but my main comment for now is to add enough test cases that cover this new change.

1. We are missing test cases for replacement context. We should check that during replacement, we don't havoc more than (the struct field) we are supposed to havoc
2. Both test cases assign to the first field of the struct. Since we look at the offsets etc, could we make one of the test assign to the second field? I'd suggest changing `regression/contracts/assigns_enforce_structs_04/main.c` to assign to `p->y` instead.

File: src/goto-instrument/contracts/assigns.cpp (hunk 3 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/assigns.cpp]: Rebase problem. Removed it.

**Comment by feliperodri** [comment]
>     1. We are missing test cases for replacement context. We should check that during replacement, we don't havoc more than (the struct field) we are supposed to havoc

This PR is only for enforcement. We should add test cases for replacement once we update the havoc method.

>     2. Both test cases assign to the first field of the struct. Since we look at the offsets etc, could we make one of the test assign to the second field? I'd suggest changing `regression/contracts/assigns_enforce_structs_04/main.c` to assign to `p->y` instead.

I added two new test cases, could you take another look? @SaswatPadhi

File: src/goto-instrument/contracts/contracts.cpp [hunk 4]
```diff
@@ -705,7 +704,7 @@ void code_contractst::instrument_assign_statement(
     instruction_iterator->source_location));
   alias_assertion.instructions.back().source_location.set_comment(
     "Check that " + from_expr(ns, lhs.id(), lhs) + " is assignable");
-  int lines_to_iterate = alias_assertion.instructions.size();
+  size_t lines_to_iterate = alias_assertion.instructions.size();
   program.insert_before_swap(instruction_iterator, alias_assertion);
   std::advance(instruction_iterator, lines_to_iterate);
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: If you rebase your branch on to the current `develop`, you wouldn't need to keep track of `lines_to_iterate` any more. I added a helper function for this, and you would be able to do:

```c
insert_before_swap_and_advance(program, instruction_iterator, alias_assertion);
```

(same for all other occurrences of `insert_before_swap` followed by `std::advance` pattern in this file)

File: src/goto-instrument/contracts/contracts.cpp [hunk 5]
```diff
@@ -753,10 +752,10 @@ void code_contractst::instrument_call_statement(
...
       // Make freshly allocated memory assignable, if we can determine its type.
       assigns_clause_targett *new_target =
-        assigns_clause.add_pointer_target(rhs);
+        assigns_clause.add_target(dereference_exprt(rhs));
       goto_programt &pointer_capture = new_target->get_init_block();
 
-      int lines_to_iterate = pointer_capture.instructions.size();
+      size_t lines_to_iterate = pointer_capture.instructions.size();
       program.insert_before_swap(local_instruction_iterator, pointer_capture);
       std::advance(instruction_iterator, lines_to_iterate + 1);
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: If you rebase your branch on to the current `develop`, you wouldn't need to keep track of `lines_to_iterate` any more. I added a helper function for this, and you would be able to do:

```c
insert_before_swap_and_advance(program, instruction_iterator, pointer_capture);
```

I don't understand the `+ 1` on the `lines_to_iterate` though. Why is it needed?

**Review by SaswatPadhi — APPROVED** [review]
Thanks, LGTM. If you rebase on to `develop`, you could remove some repeated code. Commented below:

File: src/goto-instrument/contracts/contracts.cpp (hunk 4 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/contracts.cpp]: Great! Implemented as suggested!

File: src/goto-instrument/contracts/contracts.cpp (hunk 5 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/contracts.cpp]: `+1` wasn't necessary there, it came from this case:
```c
    // malloc statments return a void pointer, which is then cast and assigned
    // to a result variable. We iterate one line forward to grab the result of
    // the malloc once it is cast.
```
I re-wrote the code to use `insert_before_swap_and_advance`. Good catch!

Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: Awesome; thanks!

**Pull request merged**